projects/client/RabbitMQ.Client/src/client/impl/AsyncConsumerWorkService.cs

```diff
@@ -75,30 +61,26 @@ public void Start()
             public void Enqueue(Work work)
             {
                 workQueue.Enqueue(work);
-                messageArrived.TrySetResult(true);
+                semaphore.Release();
             }
 
             async Task Loop()
             {
                 while (tokenSource.IsCancellationRequested == false)
                 {
-                    Work work;
-                    while (workQueue.TryDequeue(out work))
+                    await semaphore.WaitAsync().ConfigureAwait(false);
+                    if (workQueue.TryDequeue(out Work work))
                     {
                         await work.Execute(model).ConfigureAwait(false);
                     }
-
-                    await Task.WhenAny(Task.Delay(waitTime, tokenSource.Token), messageArrived.Task).ConfigureAwait(false);
-                    messageArrived.TrySetResult(true);
-                    messageArrived = new TaskCompletionSource<bool>();
```

Comment:
There is no guarantee that Enqueue and Loop are working on the same messageArrived instances here. This is a lot easier to solve with SemaphoreSlim (Enqueue calls Release and increments it, releasing the await in Loop if it's at 0), and gets rid of the 100ms polling Task.Delay. Much cleaner and more efficient.

Reply:
Makes sense.

Reply:
The reason why I used a TCS here is because it is much faster than the semaphore version. The reason is that releasing the semaphore on the hot path vs TrySetReslut on the TCS is up to 7 to 8 times slower on NETFRAMEWORK and up to 3-4 times slower on NETCORE 3.1

### Netframework without contention

|               Method |      Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |----------:|----------:|----------:|------:|--------:|------:|------:|------:|----------:|
| TaskCompletionSource |  1.874 ns | 0.0431 ns | 0.0382 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|        SemaphoreSlim | 15.351 ns | 0.0410 ns | 0.0342 ns |  8.21 |    0.18 |     - |     - |     - |         - |

### Netframework with contention

7
|               Method |      Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |----------:|----------:|----------:|------:|--------:|------:|------:|------:|----------:|
| TaskCompletionSource |  2.137 ns | 0.0650 ns | 0.0608 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|        SemaphoreSlim | 15.211 ns | 0.2943 ns | 0.2753 ns |  7.13 |    0.29 |     - |     - |     - |         - |

### Netcore 3.1 with contention

|               Method |      Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |----------:|----------:|----------:|------:|--------:|------:|------:|------:|----------:|
| TaskCompletionSource |  4.221 ns | 0.1009 ns | 0.0944 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|        SemaphoreSlim | 13.062 ns | 0.1891 ns | 0.1769 ns |  3.10 |    0.07 |     - |     - |     - |         - |

Do you want me to use the TCS again and send a PR to fix this @michaelklishin @lukebakken ?

See https://github.com/danielmarbach/MicroBenchmarks/commit/516eac917e06f5e5bdbfa794f89eb583ed98e2a3

Reply:
Please submit a PR. Thanks for the benchmarks. I assume that the original purpose of #687 (fixing a potential deadlock) will still apply.

Reply:
My concern is @stebet 's comment:

> It also replaces AutoResetEvent and TaskCompletionSource objects with SemaphoreSlim to fix a possible race-condition that could trigger a deadlock

I don't know the specifics of the deadlock so I'm relying on your expertise!

Reply:
I think he talks about the TCS not using `TaskCreationOptions.RunContinuationsAsynchronously` which we couldn't at the time of writing this code because of the netframework dependency

Reply:
Here we go https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/762